codex-rs/login/src/lib.rs

```diff
@@ -269,6 +287,97 @@ pub fn login_with_api_key(codex_home: &Path, api_key: &str) -> std::io::Result<(
     write_auth_json(&get_auth_file(codex_home), &auth_dot_json)
 }
 
+async fn run_auto_refresh_loop(auth: CodexAuth) {
+    // Minimum sleep used when we cannot infer expiry; refresh roughly hourly.
+    let default_sleep = Duration::from_secs(55 * 60);
+    loop {
+        // Determine next refresh time from access_token exp - 5 minutes.
+        let maybe_exp;
+        {
+            let auth_json = auth.get_current_auth_json();
+            if let Some(AuthDotJson {
+                tokens: Some(tokens),
+                ..
+            }) = auth_json
+            {
+                let exp = decode_jwt_exp(&tokens.access_token);
+                maybe_exp = exp.map(|dt| dt - chrono::Duration::minutes(5));
+            } else {
+                maybe_exp = None;
+            }
+        }
+
+        let now = chrono::Utc::now();
+        let sleep_for = match maybe_exp {
+            Some(when) if when > now => (when - now).to_std().unwrap_or(Duration::from_secs(0)),
+            Some(_) => Duration::from_secs(0),
+            None => default_sleep,
+        };
```

Comment:
I think this would read cleaner if you broke out a couple of functions:
1. `time_until_expiry`
2. `expiry_sleep(time_until_expiry): Option<…>` which would do the -5 minutes and also potentially return None if it's >1 day in the future or something